doc: add basic I2P documentation (#22250)
bitcoin/bitcoin: doc/i2p.md

> # I2P SUPPORT IN BITCOIN

Reviewer: No need to use full-caps here. We don't use this in other documents, and besides, the idea of a high-level formatting language like markdown is to leave it up to the rendering engine how different styles are shown.

Author: I actually lowercased that **before** reading your comment! Also changed BITCOIN to Bitcoin Core, since this doc is specific to Bitcoin Core.

It was a copy-paste-modify from doc/tor.md which uses uppercase.

> It is possible to use Bitcoin Core to accept and make connections from/to other

Reviewer 2: ```suggestion
It is possible to use Bitcoin Core to accept and make connections to/from other
```

Author: Since the wording is "accept from" and "connect to" and since "accept" comes before "connect", I put "from" before "to". Reworded in a different way now.

## Run Bitcoin Core with an I2P router

A running I2P router (proxy) with [SAM](https://geti2p.net/en/docs/api/samv3)
enabled is required (there is an [official one](https://geti2p.net) and
[a few alternatives](https://en.wikipedia.org/wiki/I2P#Routers)). Notice the IP
address and port the SAM proxy is listening to. Usually that is
`127.0.0.1:7656`. Once it is up and running with SAM enabled, the following
Bitcoin Core options can be used to engage with it:

```
-i2psam=<ip:port>
     I2P SAM proxy to reach I2P peers and accept I2P connections (default:
     none)

-i2pacceptincoming
     If set and -i2psam is also set then incoming I2P connections are
     accepted via the SAM proxy. If this is not set but -i2psam is set
     then only outgoing connections will be made to the I2P network.
     Ignored if -i2psam is not set. Listening for incoming I2P
     connections is done through the SAM proxy, not by binding to a
     local address and port (default: 1)
```

In a typical situation, this suffices:

```
bitcoind -i2psam=127.0.0.1:7656
```

The first time Bitcoin Core connects to the I2P router its I2P address (and

> its corresponding private key) will be automatically generated and saved in a

Reviewer 2: - s/router/router,/
- s/and its/and/

Author: Done.

file named `i2p_private_key` in the Bitcoin Core data directory.

## Additional configuration options related to I2P

You may set the `-debug=i2p` config logging option to have additional
information in the debug log about your I2P configuration and connections.

It is possible to restrict outgoing connections in the usual way with
`-onlynet=i2p`. Notice that, I2P support being new (as of 2021), there are not
as many I2P peers as Tor or IP ones. So using `-onlynet=i2p` may make a node
more susceptible to
[Sybil attack](https://en.bitcoin.it/wiki/Weaknesses#Sybil_attack). Use CLI
`-addrinfo` to see the number of I2P addresses known to your node.

## I2P related information in Bitcoin Core

There are several ways to see your I2P address in Bitcoin Core:
- in the debug log (grep for `AddLocal`, the I2P address ends in `.b32.i2p`)
- in the output of RPC `getnetworkinfo` in the "localaddresses" section
- in the output of the CLI `-netinfo` peer connections dashboard

To see to which I2P peers your node is connected, use `bitcoin-cli -netinfo 4`

> or the RPC `getpeerinfo` (e.g. `bitcoin-cli getpeerinfo`).

Reviewer 2: Maybe mention that you can see which I2P addresses your node knows with RPC `getnodeaddresses 0 i2p` and (maybe, not sure) that there are I2P seed nodes to bootstrap your I2P connections in `/contrib/seeds/nodes_main.txt`

Author: Added `getnodeaddresses 0 i2p`, but not the seeds mention.

Actually, we use the seeds to fill addrman / establish a connection if we don't have any. But as far as I am aware we do this regardless of the network. Should we do that per-network? I.e. make sure we establish at least one connection to Tor if the user has bothered to setup a Tor proxy, using the seeds if necessary? Same for I2P.

Reviewer 2: This seems like a good idea. IRC discussion: https://www.erisian.com.au/bitcoin-core-dev/log-2021-06-18.html#l-289

Reviewer 2: The few times I talked people through testing an I2P service, each time they needed an I2P address to connect to or wouldn't see one.  @prayank23's review is another good example :+1:

Author: "(N being either I2P or Tor). If N is configured, make sure we have at least one connection to N (either inbound or outbound). If nobody connects to us, then try to open a connection to a peer from N (using either the seeds or addresses received via gossip)".

This will help I2P now / 22.0. But it is too late for 22.0. 

For 23.0 it may not be necessary for I2P if it gains traction quickly. I am not sure if Tor needs it at all.

Reviewer 2: If demonstrated to be an issue, it may (possibly) not be too late.

## Compatibility

Bitcoin Core uses the [SAM v3.1](https://geti2p.net/en/docs/api/samv3) protocol

> to connect to the I2P network. Any I2P router that supports it can be used.

Reviewer: Please add a link to this document in `doc/README.md` (the best place is likely where `tor.md` is mentioned) so that people can find it more easily.

Author: Done.

# I2P support in Bitcoin Core

It is possible to use Bitcoin Core to accept connections from and make
connections to other Bitcoin nodes through the
[I2P network](https://en.wikipedia.org/wiki/I2P).

> [I2P glossary](https://geti2p.net/en/about/glossary) may be useful to get

Author: Done.

started with I2P terminology.

## Run Bitcoin Core with an I2P router (proxy)

A running I2P router (proxy) with [SAM](https://geti2p.net/en/docs/api/samv3)

> enabled is required (there is an [official one](https://geti2p.net) and

Reviewer 3: This official link has one 'download' button, it will open https://geti2p.net/en/download which has options for different OS. If you try installing `i2p` for Ubuntu from here it won't work. Others also look like .jar files

[a few alternatives](https://en.wikipedia.org/wiki/I2P#Routers)). Notice the IP

> address and port the SAM proxy is listening to. Usually that is

Author: Done.

`127.0.0.1:7656`. Once it is up and running with SAM enabled, the following

> Bitcoin Core options can be used to engage with it:

Author: Reworded in another way.

```
-i2psam=<ip:port>
     I2P SAM proxy to reach I2P peers and accept I2P connections (default:
     none)

-i2pacceptincoming
     If set and -i2psam is also set then incoming I2P connections are
     accepted via the SAM proxy. If this is not set but -i2psam is set
     then only outgoing connections will be made to the I2P network.
     Ignored if -i2psam is not set. Listening for incoming I2P
     connections is done through the SAM proxy, not by binding to a
     local address and port (default: 1)
```

In a typical situation, this suffices:

```
bitcoind -i2psam=127.0.0.1:7656
```

The first time Bitcoin Core connects to the I2P router, its I2P address (and
corresponding private key) will be automatically generated and saved in a file
named `i2p_private_key` in the Bitcoin Core data directory.

## Additional configuration options related to I2P

You may set the `-debug=i2p` config logging option to have additional

> information in the debug log about your I2P configuration and connections.

Author: Done.

It is possible to restrict outgoing connections in the usual way with
`-onlynet=i2p`. Notice that, I2P support being new (as of 2021), there are not
as many I2P peers as Tor or IP ones. So using `-onlynet=i2p` may make a node

> more susceptible to

Reviewer 2: - A propos "Notice that..." and "So using...", maybe something like:

"I2P support was added to Bitcoin Core in version 22.0 and there may be fewer I2P peers than Tor ones. Therefore, using..." 

- s/attack/attacks/

Author: Changed to "attacks", but left it as is otherwise.

Reviewer 2: Definitely no need to use my suggestion, but if helpful, it's incorrect to begin a sentence with "So using"... and "Notice that" isn't great either (and, like "note that," it is unneeded and can be omitted).

Author: Actually mentioning this was Gleb's [idea](https://github.com/bitcoin/bitcoin/pull/20685#issuecomment-747959538). Poking him here to ensure that I conveyed it correctly.

cc @naumenkogs

Reviewer 4: Not a major nit, but @jonatacks suggestion is grammatically correct compared to the current sentence.

Iterating on top of jons suggestion

```
I2P support was added to Bitcoin Core in version 22.0 and there may be fewer I2P peers than on the Tor or IP networks. Therefore, using...
```

Author: Reworded. "Gramattik", what is that?

[Sybil attack](https://en.bitcoin.it/wiki/Weaknesses#Sybil_attack). Use CLI

> `-addrinfo` to see the number of I2P addresses known to your node.

Reviewer 2: maybe "Run `bitcoin-cli -addrinfo` to see..."

Author: Done.

# I2P support in Bitcoin Core

It is possible to run Bitcoin Core as an
[I2P (Invisible Internet Project)](https://en.wikipedia.org/wiki/I2P)
service and connect to such services.

> This [I2P glossary](https://geti2p.net/en/about/glossary) may be useful to get

Reviewer 2: ```suggestion
This [glossary](https://geti2p.net/en/about/glossary) may be useful to get
```

Reviewer 4: agree this could just be `glossary`

Author: Done.

```

The first time Bitcoin Core connects to the I2P router, its I2P address (and
corresponding private key) will be automatically generated and saved in a file
named `i2p_private_key` in the Bitcoin Core data directory.

## Additional configuration options related to I2P

You may set the `debug=i2p` config logging option to have additional
information in the debug log about your I2P configuration and connections. Run
`bitcoin-cli help logging` for more information.

It is possible to restrict outgoing connections in the usual way with
`onlynet=i2p`. Notice that, I2P support being new (as of 2021), there are not
as many I2P peers as Tor or IP ones. So using `onlynet=i2p` may make a node
more susceptible to
[Sybil attacks](https://en.bitcoin.it/wiki/Weaknesses#Sybil_attack). Use
`bitcoin-cli -addrinfo` to see the number of I2P addresses known to your node.

## I2P related information in Bitcoin Core

There are several ways to see your I2P address in Bitcoin Core:
- in the debug log (grep for `AddLocal`, the I2P address ends in `.b32.i2p`)
- in the output of RPC `getnetworkinfo` in the "localaddresses" section
- in the output of `bitcoin-cli -netinfo` peer connections dashboard

> To see to which I2P peers your node is connected, use `bitcoin-cli -netinfo 4`

Reviewer 2: ```suggestion
To see which I2P peers your node is connected to, use `bitcoin-cli -netinfo 4`.
```

Author: Done.

or the RPC `getpeerinfo` (e.g. `bitcoin-cli getpeerinfo`).

> To see which I2P addresses your node knows, use the RPC

Reviewer 2: ```suggestion
To see which I2P addresses your node knows, run RPC
```

Author: No

Reviewer 2: "run" because it's not just the name of an RPC but a call with specific arguments. But definitely a pico-nit.
